src/backend/distributed/commands/function.c

```diff
@@ -615,3 +632,492 @@ TriggerSyncMetadataToPrimaryNodes(void)
+
+/*
+ * PlanDropFunctionStmt gets called during the planning phase of a DROP FUNCTION statement
+ * and returns a list of DDLJob's that will drop any distributed functions from the
+ * workers.
+ *
+ * The DropStmt could have multiple objects to drop, the list of objects will be filtered
+ * to only keep the distributed functions for deletion on the workers. Non-distributed
+ * functions will still be dropped locally but not on the workers.
+ */
+List *
+PlanDropFunctionStmt(DropStmt *stmt, const char *queryString)
+{
+	List *oldObjectWithArgsList = stmt->objects;
+	List *distributedObjectWithArgsList = NIL;
+	List *distributedFunctionAddresses = NIL;
+	ListCell *addressCell = NULL;
+	const char *dropStmtSql = NULL;
+	List *commands = NULL;
+	ListCell *objectWithArgsListCell = NULL;
+
+	AssertIsFunctionOrProcedure(stmt->removeType);
+
+	if (creating_extension)
+	{
+		/*
+		 * extensions should be created separately on the workers, types cascading from an
+		 * extension should therefor not be propagated here.
+		 */
+		return NIL;
+	}
+
+	if (!EnableDependencyCreation)
+	{
+		/*
+		 * we are configured to disable object propagation, should not propagate anything
+		 */
+		return NIL;
+	}
+
+
+	/*
+	 * Our statements need to be fully qualified so we can drop them from the right schema
+	 * on the workers
+	 */
+	QualifyTreeNode((Node *) stmt);
+
+	/*
+	 * iterate over all functions to be dropped and filter to keep only distributed
+	 * functions.
+	 */
+	foreach(objectWithArgsListCell, oldObjectWithArgsList)
+	{
+		ObjectWithArgs *func = NULL;
+		ObjectAddress *address = NULL;
+
+		func = castNode(ObjectWithArgs, lfirst(objectWithArgsListCell));
+		address = FunctionToObjectAddress(stmt->removeType, func, stmt->missing_ok);
+
+		if (!IsObjectDistributed(address))
+		{
+			continue;
+		}
+
+		/* collect information for all distributed functions */
+		distributedFunctionAddresses = lappend(distributedFunctionAddresses, address);
+		distributedObjectWithArgsList = lappend(distributedObjectWithArgsList, func);
+	}
+
+	if (list_length(distributedObjectWithArgsList) <= 0)
+	{
+		/* no distributed functions to drop */
+		return NIL;
+	}
+
+	/*
+	 * managing types can only be done on the coordinator if ddl propagation is on. when
+	 * it is off we will never get here. MX workers don't have a notion of distributed
+	 * types, so we block the call.
+	 */
+	EnsureCoordinator();
+
+	/* remove the entries for the distributed objects on dropping */
+	foreach(addressCell, distributedFunctionAddresses)
+	{
+		ObjectAddress *address = (ObjectAddress *) lfirst(addressCell);
+		UnmarkObjectDistributed(address);
+	}
+
+	/*
+	 * Swap the list of objects before deparsing and restore the old list after. This
+	 * ensures we only have distributed functions in the deparsed drop statement.
+	 */
+	oldObjectWithArgsList = stmt->objects;
+	stmt->objects = distributedObjectWithArgsList;
+	dropStmtSql = DeparseTreeNode((Node *) stmt);
+	stmt->objects = oldObjectWithArgsList;
```

Comment:
Guess you refer to https://github.com/citusdata/citus/pull/2991#discussion_r326899326

I thought you were pointing out that the backup of the pointer was taken at the top of the function and restored here. I have now added a assignment of the old value right above this block.

Your comment here suggests this is not what you referred to, can you elaborate on what you want to see changed?

Reply:
I meant that instead of modifying the original statement and having to change it back we could instead modify a throwaway copy (with a name that identifies its purpose).

Reply:
check, will do so!